File: AOJ/0385/0385.py
INF = 2000000000
diff = 0

def update(i, v):
    global diff
    if A[i - 1] > A[i]: diff -= 1
    if A[i] > A[i + 1]: diff -= 1
    A[i] = v
    if A[i - 1] > A[i]: diff += 1
    if A[i] > A[i + 1]: diff += 1

def sort():
    global diff
    for i in range(1, N + 1):
        if  A[i - 1] > A[i]: diff += 1

    if diff == 0: return 0
    Q = int(input())
    for i in range(1, Q + 1):
        x, y = map(int, input().split())
        v1, v2 = A[x], A[y]
        update(x, v2)
        update(y, v1)
        if diff == 0:  return i

    return -1

N = int(input())
A = list(map(int, input().split()))
A.insert(0, -INF)
A.append(INF)
print(sort())

File: AOJ/0385/0385.c
#include <stdio.h>
#define INF 2000000000
int N, A[300002], diff;

void update(int i, int v){
  if ( A[i - 1] > A[i] ) diff--;
  if ( A[i] > A[i + 1] ) diff--;
  A[i] = v;
  if ( A[i - 1] > A[i] ) diff++;
  if ( A[i] > A[i + 1] ) diff++;
}

int sort(){
  int Q, x, y, v1, v2, i;
  diff = 0;
  for ( i = 1; i <= N; i++ ) 
    if ( A[i - 1] > A[i] ) diff++;

  if ( diff == 0 ) return 0;

  scanf("%d", &Q);
  for ( i = 1; i <= Q; i++ ){
    scanf("%d %d", &x, &y);
    v1 = A[x];
    v2 = A[y];
    update(x, v2);
    update(y, v1);
    if ( diff == 0 ) return i;
  }

  return -1;
}

int main() {
  int i;
  scanf("%d", &N);
  for ( i = 1; i <= N; i++ ) scanf("%d", &A[i]);
  A[0] = -INF;
  A[N + 1] = INF;
  printf("%d\n", sort());
  return 0;
}

